### PR TITLE
feat: organizations: import csv

### LINKED ISSUE #19765: [CAL-5265] organizations: import csv (same format as export)
<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-5265](https://linear.app/calcom/issue/CAL-5265/organizations-import-csv-same-format-as-export)</sub>

### PR DESCRIPTION
**Pull request by abhijeetsingh-22**
# Update CSV Import to Handle Existing Members and Pending Invitations

## Summary

This PR extends the CSV organization member import functionality to update existing members and users with pending invitations instead of ignoring them. Previously, when importing via CSV, users who were already organization members or had pending invitations would be filtered out and skipped. Now these users get their attributes (roles, team memberships) updated based on the CSV data.

**Key Changes:**
- Added `handleExistingMemberUpdates` function to process existing members with `USER_ALREADY_INVITED_OR_MEMBER` status
- Modified `inviteMember.handler.ts` to separately handle existing users and update their roles
- Updated return value to include count of updated existing users in the success message
- Maintains backward compatibility for normal invitation flow

**Link to Devin run:** https://app.devin.ai/sessions/7bd7fc23b1764523b36de023a3ff904a  
**Requested by:** Devanshu Sharma

## Review & Testing Checklist for Human

**⚠️ HIGH PRIORITY - Core functionality not tested end-to-end:**

- [ ] **Test CSV import with existing organization members** - Upload a CSV containing users who are already members with different roles. Verify their roles get updated correctly in the database and UI.
- [ ] **Test CSV import with users who have pending invitations** - Upload a CSV containing users who have pending invites. Verify they get processed and updated appropriately.
- [ ] **Verify normal invitation flow still works** - Test standard CSV import with completely new users to ensure existing functionality isn't broken.
- [ ] **Check database integrity** - Verify that role updates don't cause any data consistency issues or violate business rules (e.g., organization must have at least one owner).
- [ ] **Test success message accuracy** - Confirm that the returned `numUsersInvited` count includes updated existing users and displays correctly in the UI toast.

**Recommended Test Plan:**
1. Create a test organization with some existing members
2. Create a CSV with mix of: new users, existing members (with role changes), and users with pending invitations
3. Import the CSV and verify each category gets handled correctly
4. Check the database directly to confirm role updates were applied
5. Test edge cases like trying to update an owner's role or updating to invalid roles

---

### Diagram

```mermaid
%%{ init : { "theme" : "default" }}%%
graph TD
    CSV["CSV Upload<br/>ImportMembersModal"]:::context
    Handler["inviteMember.handler.ts<br/>Main invitation logic"]:::major-edit
    Utils["utils.ts<br/>Processing functions"]:::major-edit
    
    CSV --> Handler
    Handler --> Utils
    
    Handler --> NewUsers["handleNewUsersInvites<br/>(existing function)"]:::context
    Handler --> ExistingInvites["handleExistingUsersInvites<br/>(existing function)"]:::context
    Handler --> ExistingMembers["handleExistingMemberUpdates<br/>(NEW function)"]:::major-edit
    
    
    ExistingMembers --> DB["Database<br/>prisma.membership.update"]:::context
    
    subgraph Legend
        L1[Major Edit]:::major-edit
        L2[Minor Edit]:::minor-edit  
        L3[Context/No Edit]:::context
    end
    
    classDef major-edit fill:#90EE90
    classDef minor-edit fill:#87CEEB
    classDef context fill:#FFFFFF
```

### Notes

- The core logic assumes that users with `USER_ALREADY_INVITED_OR_MEMBER` status have a `newRole` property set from the CSV parsing
- Database updates use `prisma.membership.update` with `userId_teamId` composite key
- Only role updates are currently implemented - could be extended for other attributes like team assignments
- Changes maintain the existing invitation flow structure and error handling patterns
- **Critical:** This functionality was not tested locally due to environment constraints - thorough manual testing is essential before merge

### CODE REVIEW

File: packages/features/users/components/UserTable/ImportMembersModal.tsx [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+                        if (val) onChange(val.value);
+                      }}
+                    />
+                    <p className="text-subtle mt-2 text-sm">{t("role_applied_if_not_specified_in_csv")}</p>
+                  </div>
+                )}
+              />
+            )}
+
+            {parsedUsers.length > 0 && (
```
Comment (TusharBhatt1) [inline packages/features/users/components/UserTable/ImportMembersModal.tsx:193]: Why are we using && && and not ternary operator (? :)

File: packages/features/users/components/UserTable/ImportMembersModal.tsx [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+                      ))}
+                    </tbody>
+                  </table>
+                </div>
+              </div>
+            )}
+
+            <div className="bg-muted mt-4 rounded-md p-4">
+              <div className="flex items-center">
+                <Icon name="info" className="text-subtle mr-2 h-4 w-4" />
```
Comment (TusharBhatt1) [inline packages/features/users/components/UserTable/ImportMembersModal.tsx:222]: Should be `size-4` instead of `h-4 w-4`

File: packages/features/users/components/UserTable/ImportMembersModal.tsx [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+      }
+    };
+
+    reader.readAsText(file);
+  };
+
+  const handleSubmit = () => {
+    if (parsedUsers.length === 0) {
+      return;
+    }
```
Comment (TusharBhatt1) [inline packages/features/users/components/UserTable/ImportMembersModal.tsx:139]: I guess we can safely remove it , we have already made the `submit` button disabled when this condition meets

**Review by TusharBhatt1 — CHANGES_REQUESTED** [review]
We should show ideal warnings for two cases :
- For invalid role names (and other attributes)
- For invalid file type (.csv)

**Review by PeerRich — COMMENTED** [review]
we also need to import attributes

best is that the "export" gives you a CSV with all attributes as a column so people can export & import

**Comment by TusharBhatt1** [comment]
> we also need to import attributes
> 
> best is that the "export" gives you a CSV with all attributes as a column so people can export & import

Yeah , so it's not only about `Members` and `Roles` , there can be other attributes , can you add the logic for the same

**Comment by abhijeetsingh-22** [comment]
> we also need to import attributes
> 
> best is that the "export" gives you a CSV with all attributes as a column so people can export & import

Sure I'll add importing attributes as well.

**Comment by eunjae-lee** [comment]
This PR looks pretty good to me. Feel free to ping me when you add the logic so that I can review it too.
Btw, I have [this PR](https://github.com/calcom/cal.com/pull/20066) going on, which is going to change location of some buttons. Depending on when it gets merged, it may affect your PR.

**Comment by abhijeetsingh-22** [comment]
@PeerRich @eunjae-lee 
I have a question about the behavior of import:  

- If a user is already a member of an organization and their email appears in the imported CSV, should we ignore it or update their attributes?  
- Similarly, if a user already has a pending invitation, should we update their attributes or ignore them?

**Comment by eunjae-lee** [comment]
@abhijeetsingh-22 yes please update attributes for both cases 🙏🏻

File: packages/features/users/components/UserTable/ImportMembersModal.tsx [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+interface FormValues {
+  defaultRole: MembershipRole;
+}
+
+export function ImportMembersModal(props: Props) {
+  const { data: session } = useSession();
+  const { data: platformUser } = usePlatformMe();
+  const utils = trpc.useUtils();
+  const { t, i18n } = useLocale();
```
Comment (Devanshusharma2005) [inline packages/features/users/components/UserTable/ImportMembersModal.tsx:44]: we use only t object for locales here.

**Review by Devanshusharma2005 — COMMENTED** [review]
Could you please add a loom to show the changes.

**Comment by Devanshusharma2005** [comment]
Closing due to inactivity. Feel free to reopen if you wanted to continue

**Comment by abhijeetsingh-22** [comment]
Hey team, apologies for the delay on this. The PR went stale on my side.
I’ve now addressed all the requested feedbacks. Would you prefer reopening this PR, or should I open a new one ?

Happy to go with whatever’s easiest for review. 
Thanks!
